### PR TITLE
[Verifier] Check function attributes related to branch protection (NFC)

### PR DESCRIPTION
**Pull request by momchil-velikov**

### CODE REVIEW

File: llvm/lib/IR/Verifier.cpp [hunk 1]
```diff
@@ -2231,6 +2231,27 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
   checkUnsignedBaseTenFuncAttr(Attrs, "patchable-function-prefix", V);
   checkUnsignedBaseTenFuncAttr(Attrs, "patchable-function-entry", V);
   checkUnsignedBaseTenFuncAttr(Attrs, "warn-stack-size", V);
+
+  if (Attrs.hasFnAttr("sign-return-adress")) {
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:2235]: all typo'd adress

File: llvm/lib/IR/Verifier.cpp [hunk 2]
```diff
@@ -2231,6 +2231,27 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
...
+    StringRef S = Attrs.getFnAttr("sign-return-adress").getValueAsString();
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:2236]: Can do getFnAttr and check if it's valid rather than looking it up twice

File: llvm/lib/IR/Verifier.cpp [hunk 3]
```diff
@@ -2231,6 +2231,27 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
...
+    if (S != "none" && S != "all" && S != "non-leaf")
+      CheckFailed("invalid value for 'sign-return-adress' attribute: " + S, V);
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:2238]: Typo adres

**Review by arsenm — CHANGES_REQUESTED** [review]
Typos for the attribute name and also needs tests in test/Verifier

File: llvm/lib/IR/Verifier.cpp (hunk 1 above)
Comment (momchil-velikov) [reply llvm/lib/IR/Verifier.cpp]: Fixed

File: llvm/lib/IR/Verifier.cpp (hunk 2 above)
Comment (momchil-velikov) [reply llvm/lib/IR/Verifier.cpp]: Done

File: llvm/lib/IR/Verifier.cpp (hunk 3 above)
Comment (momchil-velikov) [reply llvm/lib/IR/Verifier.cpp]: Fixed

**Comment by momchil-velikov** [comment]
> Typos for the attribute name and also needs tests in test/Verifier

Ah, missed this comment, will add tests.

File: llvm/lib/IR/Verifier.cpp [hunk 4]
```diff
@@ -2231,6 +2231,26 @@ void Verifier::verifyFunctionAttrs(FunctionType *FT, AttributeList Attrs,
...
+
+  if (auto A = Attrs.getFnAttr("sign-return-address"); A.isValid()) {
+    StringRef S = A.getValueAsString();
+    if (S != "none" && S != "all" && S != "non-leaf")
+      CheckFailed("invalid value for 'sign-return-address' attribute: " + S, V);
+  }
+
+  if (auto A = Attrs.getFnAttr("sign-return-address-key"); A.isValid()) {
+    StringRef S = A.getValueAsString();
+    if (!S.equals_insensitive("a_key") && !S.equals_insensitive("b_key"))
```
Comment (arsenm) [inline llvm/lib/IR/Verifier.cpp:2243]: should this really be case insensitive?

Comment (momchil-velikov) [reply llvm/lib/IR/Verifier.cpp:2243]: This matches the assertion in `AArch64MachineFunctionInfo.cpp`. Although clang/LLVM are inconsistent, some places use `==` some `equals_insensitive`.

Comment (momchil-velikov) [reply llvm/lib/IR/Verifier.cpp:2243]: I intend to change all the places to expect lower case only, for the three branch protection attributes.

**Comment by momchil-velikov** [comment]
Ping?

**Comment by momchil-velikov** [comment]
Ping?

**Review by arsenm — APPROVED** [review]

**Pull request merged**